heron/stmgr/src/cpp/manager/instance-server.cpp

```diff
@@ -133,49 +135,50 @@ InstanceServer::~InstanceServer() {
   Stop();
   // Unregister and delete the metrics.
   for (auto immIter = instance_metric_map_.begin();
-      immIter != instance_metric_map_.end(); ++immIter) {
+            immIter != instance_metric_map_.end();) {
     sp_string instance_id = immIter->first;
+
     for (auto iter = instance_info_.begin(); iter != instance_info_.end(); ++iter) {
       if (iter->second->instance_->instance_id() != instance_id) continue;
       InstanceData* data = iter->second;
       Connection* iConn = data->conn_;
       if (!iConn) break;
       sp_string metric_name = MakeBackPressureCompIdMetricName(instance_id);
       metrics_manager_client_->unregister_metric(metric_name);
-      delete immIter->second;
     }
+
+    immIter = instance_metric_map_.erase(immIter);
```

Comment:
maybe put this line in the for loop? I am afraid someday I cant help adding the "missing" immIter++. :)

Reply:
@nwangtw Actually, it is inside for-loop. See, there are two nested for loops here and we call `instance_metric_map_.erase(immIter)` in the first loop (because the outer loop is traversing `instance_metric_map_`):

```
  for (auto immIter = instance_metric_map_.begin();
            immIter != instance_metric_map_.end();) {
    sp_string instance_id = immIter->first;

    for (auto iter = instance_info_.begin(); iter != instance_info_.end(); ++iter) {
      if (iter->second->instance_->instance_id() != instance_id) continue;
      InstanceData* data = iter->second;
      Connection* iConn = data->conn_;
      if (!iConn) break;
      sp_string metric_name = MakeBackPressureCompIdMetricName(instance_id);
      metrics_manager_client_->unregister_metric(metric_name);
    }

    immIter = instance_metric_map_.erase(immIter);
  }
```

Reply:
Yeah. I am just wondering if 

```
for (auto immIter = instance_metric_map_.begin();
       immIter != instance_metric_map_.end();
       immIter = instance_metric_map_.erase(immIter)) {
....
}

```
is more readable.

Reply:
@nwangtw  updated